### PR TITLE
[RSDK-8963] add support for generic C movement sensor (readings only)

### PR DESCRIPTION
**Pull request by gvaradarajan**

### CODE REVIEW

File: micro-rdk-ffi/src/ffi/runtime.rs [hunk 1]
```diff
@@ -145,47 +245,47 @@ pub unsafe extern "C" fn viam_server_register_c_generic_sensor(
...
+///      "name": "sensor1",
+///      "namespace": "rdk",
+///      "type": "movement_sensor",
+///      "model": "my_sensor",
+///    }`
+///
+/// Sensor specific data structure to be used in by the readings callback can be written to out
+/// returns VIAM_OK on success
+/// # Safety
+/// `ctx`, `model` must be valid pointers
```
Comment (npmenard) [inline micro-rdk-ffi/src/ffi/runtime.rs:273]: add sensor

Comment (npmenard) [inline micro-rdk-ffi/src/ffi/runtime.rs:273]: add sensor

File: micro-rdk-ffi/src/ffi/runtime.rs [hunk 2]
```diff
@@ -124,6 +126,104 @@ pub unsafe extern "C" fn viam_server_set_provisioning_fragment(
...
+    ) -> Result<Self, viam_code> {
+        let name = if let Ok(s) = unsafe { CStr::from_ptr(model) }.to_str() {
+            s
+        } else {
+            return Err(viam_code::VIAM_INVALID_ARG);
+        };
+
+        // Because registry expects a &'static str for its key, we have to copy the name passed
+        // as an argument and leak it so it remains valid for the duration of the program.
+        let name: &'static str = Box::leak(name.to_owned().into_boxed_str());
```
Comment (npmenard) [inline micro-rdk-ffi/src/ffi/runtime.rs:173]: would be nice to change in to a String so we don't leak memory anymore

File: micro-rdk-ffi/src/ffi/runtime.rs [hunk 3]
```diff
@@ -145,47 +245,47 @@ pub unsafe extern "C" fn viam_server_register_c_generic_sensor(
     model: *const c_char,
     sensor: *mut generic_c_sensor_config,
 ) -> viam_code {
-    if ctx.is_null() || model.is_null() {
+    if ctx.is_null() || model.is_null() || sensor.is_null() {
```
Comment (npmenard) [inline micro-rdk-ffi/src/ffi/runtime.rs:248]: update safety section too

File: micro-rdk-ffi/src/ffi/runtime.rs [hunk 4]
```diff
@@ -124,6 +126,104 @@ pub unsafe extern "C" fn viam_server_set_provisioning_fragment(
...
+type MovementSensorConstructor =
+    Box<dyn Fn(ConfigType, Vec<Dependency>) -> Result<MovementSensorType, SensorError>>;
+
+pub(crate) enum ConstructorType {
+    Sensor(&'static str, SensorConstructor),
+    MovementSensor(&'static str, MovementSensorConstructor),
+}
+
+impl ConstructorType {
+    pub(crate) fn register(self, ctx: &mut viam_server_context) -> viam_code {
```
Comment (npmenard) [inline micro-rdk-ffi/src/ffi/runtime.rs:141]: might consider to call register as part of  from_sensor / from_movement_sensor

File: micro-rdk-ffi/src/ffi/runtime.rs [hunk 5]
```diff
@@ -145,47 +245,47 @@ pub unsafe extern "C" fn viam_server_register_c_generic_sensor(
...
+    sensor: *mut generic_c_movement_sensor_config,
+) -> viam_code {
+    if ctx.is_null() || model.is_null() || sensor.is_null() {
+        return viam_code::VIAM_INVALID_ARG;
     }
+    let ctx = unsafe { &mut *ctx };
 
-    viam_code::VIAM_OK
+    match ConstructorType::from_movement_sensor(sensor, model) {
+        Ok(constructor) => constructor.register(ctx),
```
Comment (npmenard) [inline micro-rdk-ffi/src/ffi/runtime.rs:286]: If i am not mistaken i think this path can result in a error. If you call register as part of from_movement_sensor you can only have one path for errors which is a bit more readable IMO

File: micro-rdk-ffi/src/ffi/config.rs [hunk 6]
```diff
@@ -1,17 +1,44 @@
...
+}
+
 #[allow(non_camel_case_types)]
 pub struct config_context<'a> {
     pub(crate) cfg: ConfigType<'a>,
 }
 
+pub(crate) trait GenericCResourceConfig {
+    fn get_user_data_and_config_callback(&mut self) -> (*mut c_void, config_callback);
+    fn configure(&mut self, mut config: config_context) -> Result<*mut c_void, i32> {
```
Comment (npmenard) [inline micro-rdk-ffi/src/ffi/config.rs:30]: if you are not planning to let user modify if then not sure it should be in a trait

**Review by npmenard — CHANGES_REQUESTED** [review]
Done with first pass

**Comment by gvaradarajan** [comment]
@npmenard should I close this PR given recent events?

**Comment by gvaradarajan** [comment]
we're not doing this for now

**Pull request closed without merging**